django/db/backends/utils.py

```diff
@@ -114,6 +114,75 @@ def _executemany(self, sql, param_list, *ignored_wrapper_args):
             return self.cursor.executemany(sql, param_list)
 
 
+class AsyncCursorCtx:
+    """
+    Asynchronous context manager to hold an async cursor.
+    """
+
+    def __init__(self, db, name=None):
+        self.db = db
+        self.name = name
+        self.wrap_database_errors = self.db.wrap_database_errors
+
+    async def __aenter__(self):
+        await self.db.aclose_if_health_check_failed()
+        await self.db.aensure_connection()
+        self.wrap_database_errors.__enter__()
+        return self.db._aprepare_cursor(self.db.create_async_cursor(self.name))
+
+    async def __aexit__(self, type, value, traceback):
+        self.wrap_database_errors.__exit__(type, value, traceback)
+
+
+class AsyncCursorWrapper(CursorWrapper):
+    async def _execute(self, sql, params, *ignored_wrapper_args):
+        # Raise a warning during app initialization (stored_app_configs is only
+        # ever set during testing).
+        if not apps.ready and not apps.stored_app_configs:
+            warnings.warn(self.APPS_NOT_READY_WARNING_MSG, category=RuntimeWarning)
+        self.db.validate_no_broken_transaction()
+        with self.db.wrap_database_errors:
+            if params is None:
+                # params default might be backend specific.
+                return await self.cursor.execute(sql)
+            else:
+                return await self.cursor.execute(sql, params)
+
+    async def _execute_with_wrappers(self, sql, params, many, executor):
+        context = {"connection": self.db, "cursor": self}
+        for wrapper in reversed(self.db.execute_wrappers):
+            executor = functools.partial(wrapper, executor)
+        return await executor(sql, params, many, context)
+
+    async def execute(self, sql, params=None):
```

Comment:
Same comment as below, I really think we should avoid having async-arity-different subclassing here. 

It's much easier to debug mistaken method calls if the functions are different (if not you end up with "coroutine is not a list"-style errors _way later_ down the stack).

Reply:
There's some messiness because the underlying cursor uses a method called `execute`. One thing I regret not doing in my branch is just wrapping the async cursor in a way that _doesn't_ proxy to the underlying cursor. Had to waste a lot of time chasing down incorrect calls

Reply:
In that case, since there's no name conflict anymore, should we have one `Cursor` class with both `.execute()` and `.aexecute()`